### DISCUSSION TITLE
Making structured mesh (with brick elements) using a .geo file

### DISCUSSION THREAD

**Discussion by MohammadJoudivand** [mtex-toolbox/mtex#538] [Ask Anything]
Dear Participants
Are you able to kindly let me know how can I create a structured mesh (with brick elements) in a .geo model file that is generated using "savegeo(G,'DP600_meshed.geo')" function of the GMSH toolbox?
To be mentioned my model consists several grains which are already converted to square shape (from their primary hexagonal form).
Sincerely yours
Mohammad Hasan Joudivand

**Comment by aaditya8959** [comment]
Hi,
  The information in this [link](https://openfoamwiki.net/index.php/2D_Mesh_Tutorial_using_GMSH) might be of use to you. Section 1.3 covers an example to generate a structured mesh for a 3D object in gmsh, which is basically affected by making certain modifications to the .geo file.

Best,
Aaditya

**Reply by MohammadJoudivand** [reply]
Dear Aaditya
Thanks for your kind attention,
Actually my .geo file is extracted from an experimental EBSD file that is processed using MTEX and finally saved through  "savegeo(G,'filename.geo')" function of GMSH.
There are several grains which own nonuniform shapes (but all of the are squared before), it means I am dealing with tens of surfaces in this case.
If I only take the corners of my rectangular model and define a surface that covers and overlaps with all of the surfaces of the grain tiles, it will be non-sense because I am trying to construct my mesh based on the experimental data.
Do you know what should I do?
Please advise.
Sincerely
Mohammad

**Comment by aaditya8959** [comment]
Hi Mohammad,
   So let me get this right for my understanding. You currently have a .geo file which when visualized using gmsh, has a structured 2D mesh with square elements(obtained from an initial hexagonal arrangement), and you require a way to generate 3D mesh with 1 layer in the 3rd direction obtained by extruding the 2D mesh. If my understanding is correct, then you can probably do it in 2 ways. 

1) Extrude all surfaces in your 2D mesh in the third direction. To do that within a .geo file you might have to use some scripting alternative to generate a new .geo file with the new set of commands to extrude all the surfaces in the 2D mesh(for how this is done for a single surface look at Section 1.1, Pointer 7). There might also be a way to do this solely using gmsh, but you would have to dig deeper into the gmsh reference manual. Perhaps you can specify multiple surface indices as a loop or using some other syntax in gmsh.

2) The other alternative is to postprocess the .msh file obtained for the structure 2D mesh. In that case you just need to realize that if you wish to generate a 3D structured grid with 1 element in the 3rd direction(out of the EBSD plane) then you can use the same element numbers as the 2D mesh, but introduce extra node numbers and define new elements by redefining the arrangement with the new node numbers. You can use MATLAB for generating the new .msh file from the original one.

Eg - Suppose you have the following .msh file for a 2D structured grid.

$MeshFormat
2.2 0 8
$EndMeshFormat
$Nodes
4
1 0 0 0 
2 1 0 0 
3 1 1 0
4 0 1 0
$EndNodes
$Elements
1 3 2 0 11 1 2 3 4
$EndElements

Then the 3D version of this mesh extruded in the 3rd direction by one element should be 

$MeshFormat
2.2 0 8
$EndMeshFormat
$Nodes
8
1 0 0 0 
2 1 0 0 
3 1 1 0
4 0 1 0
5 0 0 1 
6 1 0 1 
7 1 1 1
8 0 1 1
$EndNodes
$Elements
1 5 2 0 11 1 2 3 4 5 6 7 8 
$EndElements

Notice that since there were 4 nodes in the original mesh on one face, there are now 8 nodes for the 3D grid where for each node in the 2D mesh there is a corresponding node translated in the 3rd direction which differs in the value of that coordinate. Also notice that the element type has changed from 3(quadrangle) to 5(hexahedron) and that there are 8 nodes mentioned in the rearrangement as opposed to 4. The number of square elements remains the same as the number of hexahedral elements since for a single element in the 3rd direction, for each square there is a corresponding cube. 

I would prefer the second alternative since all we need to do is work with numbers and it should be quick to automate this in MATLAB. I hope the above helps.

Best,
Aaditya

**Reply by MohammadJoudivand** [reply]
Dear Aadita
Thank you very much for your very detailed answer.
The link was very useful and I applied it in my model, but now I have encountered a problem.
Actually my model consists of several grains which are not necessarily 4-corner ones. It means that grains are tiled beside each other to form a rectangular side. Some of them have 4 corners and some others have more than 4 corners (5 (L shaped), 6 or even more) but the corners are configured in right angle (90 degrees).
Now, when I apply the "Transfinite surface" command I get an error message saying that some of the surfaces have more than 4 corners, this leads the mesh to remain unstructured in that tiles.
Do you have any suggestion in this regard?
Sincerely yours
Mohammad

**Reply by aaditya8959** [reply]
Hi Mohammad,
    Can you share an image? I am unable to visualize what you mentioned regarding more than 4 corners for some of the grains.

Best,
Aaditya

**Reply by MohammadJoudivand** [reply]
Dear Aaditya
Please find the images as attached, 
Thank you
Mohammad
![Fig1](https://user-images.githubusercontent.com/73221138/96831980-adbc5800-1446-11eb-81dc-bc2c6e581a92.jpg)
![1](https://user-images.githubusercontent.com/73221138/96832090-d9d7d900-1446-11eb-9fec-a1049ab929f7.jpg)

**Comment by aaditya8959** [comment]
Hi Mohammad,
   Why not treat each square pixel separately since they are anyways going to be independent entities later, and you have all the microstructural information associated with each pixel(grainID, orientations, etc)? I don't see any advantage in the microstructure being defined in the manner that is visible in the image, from a simulation standpoint. You can just generate a structured mesh separately(independent of the EBSD map) and then assign properties to the elements of the mesh based on the EBSD map information.

Best,
Aaditya

**Reply by MohammadJoudivand** [reply]
Dear Aaditya
In that case it would be very complicated and tough to implement since there are several modeling cases I am supposed to investigate. Furthermore I may be required to use finer mesh or change the tolerance angle which separates grains, then I need to put a lot of time for each case.
Sincerely 
Mohammad

**Reply by aaditya8959** [reply]
Hi Mohammad,
     If you don't mind, can you share an example .geo file? I would like try it from my end.

Best,
Aaditya

**Reply by MohammadJoudivand** [reply]
Yes for sure
Please find it here
I changed its format to .txt because in .geo it was not accepted to be uploaded.
Please change it to .geo
Thank you
Mohammad

[DP600_meshed.txt](https://github.com/mtex-toolbox/mtex/files/5420738/DP600_meshed.txt)

**Reply by aaditya8959** [reply]
Hi Mohammad,
   Look at the comments in lines 43-47 [here](https://fossies.org/linux/gmsh/tutorial/t6.geo). That might help. Even if you are successful in generating a mesh without the errors of more than 4 corners, I am not sure you will end up with a structured mesh like you wish. I would still lean towards automating the process of generating the .msh file using something like MATLAB which gives you a lot more control, even though it might be a little bit more work on your part.

Best,
Aaditya

**Reply by aaditya8959** [reply]
If you aren't too particular about structured meshes, maybe give [this](https://github.com/DorianDepriester/MTEX2Gmsh) a try.

**Reply by MohammadJoudivand** [reply]
Thank you very much Aaditya
I will try something else to solve this problem.
I have already been creating my mesh using MTEX2Gmsh but it is not properly working too.
Thanks for your kind attention and responses.
Bests
Mohammad
